components/scheduler2/generator/executors.go

```diff
@@ -0,0 +1,289 @@
+	overlapPolicy := scheduler.OverlapPolicy()
+
+	e.Logger.Debug("processTimeRange",
+		tag.NewTimeTag("start", start),
+		tag.NewTimeTag("end", end),
+		tag.NewAnyTag("overlap-policy", overlapPolicy),
+		tag.NewBoolTag("manual", false))
+
+	catchupWindow := e.catchupWindow(scheduler, tweakables)
+
+	// Peek at paused/remaining actions state and don't bother if we're not going to
+	// take an action now. (Don't count as missed catchup window either.)
+	// Skip over entire time range if paused or no actions can be taken.
+	if !scheduler.CanTakeScheduledAction(false) {
+		// use end as last action time so that we don't reprocess time spent paused
+		next, err := e.getNextTime(scheduler, end)
+		if err != nil {
+			return nil, err
+		}
+
+		return &processedTimeRangeResult{
+			NextWakeupTime: next.Next,
+			LastActionTime: end,
+			BufferedStarts: nil,
+		}, nil
+	}
+
+	lastAction := start
+	var next scheduler1.GetNextTimeResult
+	var bufferedStarts []*schedpb.BufferedStart
+	for next, err := e.getNextTime(scheduler, start); err == nil && !(next.Next.IsZero() || next.Next.After(end)); next, err = e.getNextTime(scheduler, next.Next) {
+		if scheduler.Info.UpdateTime.AsTime().After(next.Next) {
+			// We're reprocessing since the most recent event after an update. Discard actions before
+			// the update time (which was just set to "now"). This doesn't have to be guarded with
+			// hasMinVersion because this condition couldn't happen in previous versions.
+			continue
+		}
+
+		if end.Sub(next.Next) > catchupWindow {
+			e.Logger.Warn("Schedule missed catchup window", tag.NewTimeTag("now", end), tag.NewTimeTag("time", next.Next))
+			e.MetricsHandler.Counter(metrics.ScheduleMissedCatchupWindow.Name()).Record(1)
+
+			// TODO - update Info.MissedCatchupWindow
+			// s.Info.MissedCatchupWindow++
+			// or write that to the generator's persisted state?
+			continue
+		}
+
+		bufferedStarts = append(bufferedStarts, &schedpb.BufferedStart{
+			NominalTime:   timestamppb.New(next.Nominal),
+			ActualTime:    timestamppb.New(next.Next),
+			OverlapPolicy: scheduler.OverlapPolicy(),
+			Manual:        false,
+			RequestId:     common.GenerateRequestID(scheduler, "", next.Nominal, next.Next),
+		})
+		lastAction = next.Next
+
+		if limit != nil {
+			if (*limit)--; *limit <= 0 {
+				break
+			}
+		}
+	}
+
+	return &processedTimeRangeResult{
+		NextWakeupTime: next.Next,
+		LastActionTime: lastAction,
+		BufferedStarts: bufferedStarts,
+	}, nil
+}
+
+func (e taskExecutor) catchupWindow(s core.Scheduler, tweakables common.Tweakables) time.Duration {
+	cw := s.Schedule.Policies.CatchupWindow
+	if cw == nil {
+		return tweakables.DefaultCatchupWindow
+	}
+
+	return max(cw.AsDuration(), tweakables.MinCatchupWindow)
+}
+
+// Returns the next time result, or an error if the schedule cannot be compiled.
+func (e taskExecutor) getNextTime(s core.Scheduler, after time.Time) (scheduler.GetNextTimeResult, error) {
+	spec, err := s.CompiledSpec(e.SpecBuilder)
```

Comment:
just curious if these can get cached at all. it's not _that_ expensive to construct so probably not worth it, except for backfills where it's doing this in a loop during one transition